Update to .net 7

makotech222 opened the pull request:
All tests are passing. 
- Switched scriptloader to use Roslyn. 
- add pragma warning ignore around using binaryserializer, which is deprecated
- update xunit

Everything else was upgraded automatically without issue

makotech222 committed:
Update to .net 7

exectails commented:
Thank you for the pull request. It's nice to see that switching to .NET 7 was relatively straightforward. Unfortunately there's a minor issue with a straight port. I should've explained this in more detail, but I didn't expect anyone to actually go to work right away^^"

There are still Framework applications other than Melia that use Yggdrasil, and even if we switch Melia to Core, those other applications are relying on the continued support of Yggdrasil as well, and there might be no plans to switch them over. For this reason we had looked into switching to .NET Standard 2.0, for multi-framework support, or potentially cross-compile for multiple frameworks, instead of doing a straight up conversion.

Since my experience with Core and Standard are limited, perhaps you have some ideas for how to ideally go about this?

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -343,44 +407,20 @@ private Assembly Compile(IEnumerable<string> scriptFilesList)
 				{
-					foreach (CompilerError error in errors)
-					{
-						if (mapFilePaths.TryGetValue(error.FileName, out var fileName))
```
Please note that it's important for the user to know where the error happened. If the script file was run through a pre-compiler, the path to the file would be somewhere in the temp folder, which is why the file name in the error was replaced here before forwarding the exception.

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -343,44 +407,20 @@ private Assembly Compile(IEnumerable<string> scriptFilesList)
-							error.FileName = fileName;
-					}
-
-					throw new CompilerErrorException(errors);
```
It appears like you removed all forwarding of exceptions or errors and are simply swallowing them now. This is a no-go, as the user needs to know that the compilation failed and why. You can see in Melia how we use this information to print out errors in scripts.

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -201,6 +206,71 @@ protected List<string> ReadScriptList(string scriptListFile, params string[] pri
+			var rtPath = Path.GetDirectoryName(typeof(object).Assembly.Location) +
+						 Path.DirectorySeparatorChar;
+
+			AddAssemblies(
```
Please try to stick to the general style conventions found around the project, such as using `this.` to access publics and prefixing private fields with `_`.

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -201,6 +206,71 @@ protected List<string> ReadScriptList(string scriptListFile, params string[] pri
+						 Path.DirectorySeparatorChar;
+
+			AddAssemblies(
+				rtPath + "System.Private.CoreLib.dll",
```
I'd recommend refactoring this to use `Path.Combine` and to remove the manual `+ Path.DirectorySeparatorChar` above, as `Combine` will ensure that you get a valid path, regardless of different systems, frameworks, implementations, coding errors, and what have you.

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -201,6 +206,71 @@ protected List<string> ReadScriptList(string scriptListFile, params string[] pri
 			return result;
 		}
 
+		private bool AddAssembly(string assemblyDll)
```
Please include doc comments even on private methods, as providing summaries of what methods do makes it easier to follow the code without having to read every line.

exectails commented on Yggdrasil.Test/Yggdrasil.Test.csproj:
```diff
@@ -1,147 +1,26 @@
   </ItemGroup>
   <ItemGroup>
-    <None Include="packages.config" />
+    <PackageReference Include="xunit" Version="2.4.2" />
```
The xUnit update caused a bunch of compilation warnings to pop up because the new version wants things to be done differently from the old version. While warnings aren't a show-stopper, I expect Yggdrasil to compile without them. Not only because they bother me, but also because there could be some among those >100 warnings that we do care about.

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -64,8 +70,8 @@ public class ScriptLoader
 		/// scripts and no caching.
 		/// </summary>
 		/// <param name="provider"></param>
```
If you change signature methods, please update the comments as well. In this case for example, the parameter was removed and the summary should no longer speak of a "given compiler".

exectails commented on Yggdrasil.Scripting/ScriptLoader.cs:
```diff
@@ -14,7 +19,6 @@ namespace Yggdrasil.Scripting
 	/// </summary>
 	public class ScriptLoader
 	{
-		private readonly CodeDomProvider _compiler;
```
Letting the user specify the compiler allowed us to support different compilers, aside from C#. While losing this isn't the end of the world, it would be kind of nice to support it going forward. For example, in the past I also compiled VB.NET, Boo, and JScript with this code. Does the replacement compilation approach support this?

exectails commented on Yggdrasil.Scripting/Yggdrasil.Scripting.csproj:
```diff
@@ -1,62 +1,24 @@
-    <Compile Include="Properties\AssemblyInfo.cs" />
-    <Compile Include="ScriptLoader.cs" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.6.0" />
+    <PackageReference Include="Microsoft.CodeDom.Providers.DotNetCompilerPlatform" Version="4.1.0" />
```
I'm getting a warning that `Microsoft.CodeDom.Providers.DotNetCompilerPlatform` was made for .NET Framework and not Standard. Is this going to be an issue when using the library from a Core project?

exectails reviewed (changes requested):
Thank you for your efforts. I gave your changes a cursory look and noticed a couple (potential) issue. Btw, have you tested your updated version of Yggdrasil with Melia, or did you only run the tests for the moment?

makotech222 commented:
Okay just pushed a commit. Should fix all of those issues you brought up. Regarding passing in a compiler, I don 't think it really works with roslyn apis that way. It might be best to add a separate codepath for that.

exectails commented:
Thank you for the update! Unfortunately there are still some warnings and code convention violations. Btw, did you test this in combination with Melia yet by any chance to see if everything is working as expected?

makotech222 commented:
I just rebased, fixed a few compile issues in test. I'm not seeing any important warnings anymore; just a few reference conflicts, which should be fine. Any specific code convention issues?

exectails commented:
> Any specific code convention issues?

Particularly I noticed a few instances where the `var` keyword could be used, as well as private fields not being camelCase and prefixed with underscores.

> I'm not seeing any important warnings anymore

Warnings are rarely important but they still bug me^^ Sorry for the inconvenience.

makotech222 committed:
minor cleanup

makotech222 committed:
Fix to compilererror

exectails commented:
Seeing how there had been some brought up issues still open, as well as more that I discovered while testing Yggdrasil with Core and Framework applications, I decided to look into updating the libraries myself, using your PR as reference, to fix them and gain a better understanding of the process myself. The master branch is now running on .NET Standard 2.0, and Melia will soon be updated as well.

Apologies for taking so long, and sorry your efforts partially went unused. But I appreciated them nonetheless.
